pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/SignatureDeclareThrowsExceptionRule.java

```diff
@@ -39,172 +34,78 @@
 
 public class SignatureDeclareThrowsExceptionRule extends AbstractJavaRule {
 
-    private static final PropertyDescriptor<Boolean> IGNORE_JUNIT_COMPLETELY_DESCRIPTOR = booleanProperty("IgnoreJUnitCompletely").defaultValue(false).desc("Allow all methods in a JUnit testcase to throw Exceptions").build();
+    private static final PropertyDescriptor<Boolean> IGNORE_JUNIT_COMPLETELY_DESCRIPTOR =
+            booleanProperty("IgnoreJUnitCompletely").defaultValue(false)
+            .desc("Allow all methods in a JUnit3 TestCase to throw Exceptions").build();
 
     // Set to true when the class is determined to be a JUnit testcase
-    private boolean junitImported = false;
+    private boolean junit3TestClass = false;
 
     public SignatureDeclareThrowsExceptionRule() {
         definePropertyDescriptor(IGNORE_JUNIT_COMPLETELY_DESCRIPTOR);
     }
 
     @Override
-    public Object visit(ASTCompilationUnit node, Object o) {
-        junitImported = false;
-        return super.visit(node, o);
+    public void start(RuleContext ctx) {
+        super.start(ctx);
+        junit3TestClass = false;
     }
 
     @Override
     public Object visit(ASTClassOrInterfaceDeclaration node, Object data) {
-        if (junitImported) {
+        if (junit3TestClass) {
```

Comment:
this whole method is weird... It looks like the whole body is equivalent to
```java
        junit3TestClass = junit3TestClass || TestFrameworksUtil.isJUnit3Class(node);
        return super.visit(node, data);
```
Also this looks like a bug: it should be written
```java
        boolean prev = junit3TestClass;
        junit3TestClass = TestFrameworksUtil.isJUnit3Class(node);
        super.visit(node, data);
        junit3TestClass = prev;
        return null;
```

Reply:
Actually we can probably use the rulechain instead and make this rule much faster